pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+    return lowerMergeLevel == null || windowEndMs <= lowerMergeLevelWatermarkMs;
+  }
+
+  /**
+   * Get the watermark from the MergeRollupMetadata ZNode.
+   * If the znode is null, computes the watermark using the start time from segment metadata
+   */
+  private long getWatermarkMs(String offlineTableName, long minStartTimeMs, long bucketMs, String mergeLevel) {
+    MergeRollupTaskMetadata mergeRollupTaskMetadata =
+        _clusterInfoAccessor.getMinionMergeRollupTaskMetadata(offlineTableName);
+    long watermarkMs;
+    if (mergeRollupTaskMetadata == null || mergeRollupTaskMetadata.getWatermarkMap() == null
+        || mergeRollupTaskMetadata.getWatermarkMap().get(mergeLevel) == null) {
+      // No ZNode exists. Cold-start.
+      // Round off according to the bucket. This ensures we align the merged segments to proper time boundaries
+      // For example, if start time millis is 20200813T12:34:59, we want to create the merged segment for window [20200813, 20200814)
+      watermarkMs = (minStartTimeMs / bucketMs) * bucketMs;
+      updateWatermarkMs(offlineTableName, mergeLevel, watermarkMs);
+    } else {
+      watermarkMs = mergeRollupTaskMetadata.getWatermarkMap().get(mergeLevel);
+    }
+    return watermarkMs;
+  }
+
+  /**
+   * Update current watermark for mergeLevel
+   */
+  private void updateWatermarkMs(String offlineTableName, String mergeLevel, long newWatermarkMs) {
+    MergeRollupTaskMetadata mergeRollupTaskMetadata =
+        _clusterInfoAccessor.getMinionMergeRollupTaskMetadata(offlineTableName);
+    Map<String, Long> waterMarkMap;
+    if (mergeRollupTaskMetadata == null || mergeRollupTaskMetadata.getWatermarkMap() == null) {
+      waterMarkMap = new TreeMap<>();
+    } else {
+      waterMarkMap = mergeRollupTaskMetadata.getWatermarkMap();
+    }
+    Long prevWatermarkMs = waterMarkMap.put(mergeLevel, newWatermarkMs);
+    LOGGER.info("Update watermark of mergeLevel: {} for table: {} from: {} to: {}", mergeLevel, offlineTableName,
+        prevWatermarkMs, newWatermarkMs);
+    _clusterInfoAccessor.setMergeRollupTaskMetadata(new MergeRollupTaskMetadata(offlineTableName, waterMarkMap));
+  }
+
+  /**
+   * Create pinot task configs for selected segments and pconfigs
+   */
+  private List<PinotTaskConfig> createPinotTaskConfigs(List<OfflineSegmentZKMetadata> selectedSegments,
+      String offlineTableName, Map<String, String> taskConfigs, Map<String, String> mergeConfig) {
+    int maxNumRecordsPerTask = mergeConfig.get(MergeRollupTask.MAX_NUM_RECORDS_PER_TASK_KEY) != null ? Integer.parseInt(
+        mergeConfig.get(MergeRollupTask.MAX_NUM_RECORDS_PER_TASK_KEY)) : DEFAULT_MAX_NUM_RECORDS_PER_TASK;
+    int numRecordsPerTask = 0;
+    List<List<String>> segmentNamesList = new ArrayList<>();
+    List<List<String>> downloadURLsList = new ArrayList<>();
+    List<String> segmentNames = new ArrayList<>();
+    List<String> downloadURLs = new ArrayList<>();
```

Comment:
(nit) These 2 variables can be local within the for loop

Reply:
Hmm, they won't work if moved within the for loop. Did I miss something here?